internal/internal_workflow.go

```diff
@@ -1392,26 +1434,94 @@ func setQueryHandler(ctx Context, queryType string, handler interface{}) error {
 	return nil
 }
 
-func (h *queryHandler) validateHandlerFn() error {
-	fnType := reflect.TypeOf(h.fn)
+// setUpdateHandler sets update handler for a given update name.
+func setUpdateHandler(ctx Context, updateName string, handler interface{}, opts UpdateOptions) error {
+	if err := validateHandlerFn(handler); err != nil {
+		return err
+	}
+	var validateFn interface{} = func(...interface{}) error { return nil }
+	if opts.Validator != nil {
+		if err := validateValidatorFn(opts.Validator); err != nil {
+			return err
+		}
+		if err := validateSameParams(handler, opts.Validator); err != nil {
+			return err
+		}
+		validateFn = opts.Validator
+	}
+	getWorkflowEnvOptions(ctx).updateHandlers[updateName] = &updateHandler{
+		fn:            handler,
+		validateFn:    validateFn,
+		name:          updateName,
+		dataConverter: getDataConverterFromWorkflowContext(ctx),
+	}
+	return nil
+}
+
+func validateSameParams(fn1, fn2 interface{}) error {
+	fn1Type := reflect.TypeOf(fn1)
+	fn2Type := reflect.TypeOf(fn2)
+
+	for _, fnType := range [...]reflect.Type{fn1Type, fn2Type} {
```

Comment:
Meh, two if statements ain't that bad, heh

Reply:
lol fair point. I was fussing with test coverage. hazards of reviewing a wip ;-)